libazureinit/src/imds.rs

```diff
@@ -87,26 +90,65 @@ where
     }
 }
 
-pub async fn query(client: &Client) -> Result<InstanceMetadata, Error> {
+fn is_retryable_code(statuscode: StatusCode) -> bool {
+    // 400, 404, 410, 429, 500, 503
+    [
+        StatusCode::from_u16(400).unwrap(),
+        StatusCode::from_u16(404).unwrap(),
+        StatusCode::from_u16(410).unwrap(),
+        StatusCode::from_u16(429).unwrap(),
+        StatusCode::from_u16(500).unwrap(),
+        StatusCode::from_u16(503).unwrap(),
+    ]
+    .contains(&statuscode)
+}
+
+pub async fn query(
+    client: &Client,
+    retry_interval: Duration,
+    total_timeout: Duration,
+) -> Result<InstanceMetadata, Error> {
     let url = "http://169.254.169.254/metadata/instance?api-version=2021-02-01";
     let mut headers = HeaderMap::new();
+    let mut fastfail: bool = false;
 
     headers.insert("Metadata", HeaderValue::from_static("true"));
 
-    let request = client.get(url).headers(headers);
-    let response = request.send().await?;
+    let response = timeout(total_timeout, async {
+        loop {
+            if let Ok(response) = client
+                .get(url)
+                .timeout(Duration::from_secs(30))
+                .send()
+                .await
+            {
+                let statuscode = response.status();
+
+                if statuscode.is_success() && statuscode == 200 {
+                    break response;
+                }
 
-    if response.status().is_success() {
-        let imds_body = response.text().await?;
-        let metadata: InstanceMetadata = serde_json::from_str(&imds_body)?;
+                if !is_retryable_code(statuscode) {
+                    fastfail = true;
+                    break response;
```

Comment:
You could return this by doing `break (response, fastfail);` instead of needing to pre-declare a variable with `mut`. 

You could also return `response.error_for_status()` if it's not on the retry list and then unwrap that result with `?` outside the timeout block to get rid of the `if fastfail` entirely. `reqwest::Error` has a [url](https://docs.rs/reqwest/latest/reqwest/struct.Error.html#method.url) attached to it so we should be able to extract all the information we need from it.

It's all equivalent and what here now isn't difficult to understand, so it's totally up to you.

Reply:
Changed to `response.error_for_status()`.